### ISSUE TITLE
[FEATURE] Allow user-defined functions for score normalization and combination in hybrid queries

### ISSUE THREAD

**Issue by martin-gaievski** [opensearch-project/neural-search#994]
### Is your feature request related to a problem?
Currently, OpenSearch provides internal functionality for score normalization and combination in hybrid queries, as outlined in the [Normalization Processor documentation](https://opensearch.org/docs/latest/search-plugins/search-pipelines/normalization-processor/). However, there is a need to allow users to define their own custom functions for these operations instead of solely relying on OpenSearch's internal mechanisms.

The ability to define user-specific functions will provide more flexibility and control over how scores are normalized and combined, especially for advanced use cases where the built-in functionality may not suffice.

### What solution would you like?

Introduce the ability for users to define custom functions for score normalization and combination in hybrid queries. These functions could be implemented using:

- **Function Score Queries**: Use the existing OpenSearch function score query framework to define custom scoring logic, including:
  - **Weighting**: Apply custom weights to individual components of a query.
  - **Decay functions**: Implement decay functions based on relevance or other factors.
  - **Custom logic**: Create complex scoring models based on attributes, fields, or external parameters.

- **Painless Scripts**: Allow users to use OpenSearch's Painless scripting language to define custom functions for score normalization and combination. Painless is a lightweight, secure, and performant scripting language that integrates seamlessly with OpenSearch.

  The Painless scripting language can be used to define custom functions for complex logic, including:
  - **Custom arithmetic**: Implement custom arithmetic for combining or normalizing scores.
  - **Conditional scoring**: Apply conditional logic based on query results or document attributes.
  - **Advanced normalization**: Create custom normalization schemes to adjust scores based on user-defined rules.

We can go and further and try to implement the support for invoking external scripts (e.g., Python or SQL) for even more sophisticated logic, where the internal scripting options may not be sufficient. This could allow users to execute pre-defined models or complex scoring algorithms that are managed externally.

**Benefits:**

- **Flexibility**: Users can create highly customized scoring and normalization logic, tailored to their specific use case.
- **Advanced Control**: Users gain fine-grained control over how scores are combined or normalized, leading to more relevant and accurate search results.
- **Extensibility**: Users can leverage external tools (e.g., machine learning models, databases, or business rules) for scoring, creating a unified search pipeline.
- **Consistency**: Streamline the process of integrating custom scoring logic across different search types and models, ensuring uniformity in query results.

**Use Case:**

For example, a user may want to combine results from multiple models (e.g., semantic search and traditional keyword search) and apply a custom score normalization function. Using a Painless script, the user could adjust the score of each result based on a combination of the model score and some external business logic (e.g., boosting certain results based on document metadata or user preferences).

Alternatively, a user might prefer to use a Python script to implement a more complex machine learning model for score normalization, offering them the flexibility to include custom ranking logic, external data, or ML-based techniques.

**Comment by navneet1v** [comment]
@martin-gaievski I remember MLcommons has the pre and post functions that can run on the embeddings we should see how they are doing it.

**Comment by dblock** [comment]
[[Catch All Triage](https://github.com/opensearch-project/.github/blob/main/TRIAGING.md) - [1](https://github.com/dblock), [2](https://github.com/reta), [3](https://github.com/andrross/), [4](https://github.com/krisfreedain)]

**Comment by martin-gaievski** [comment]
> [@martin-gaievski](https://github.com/martin-gaievski) I remember MLcommons has the pre and post functions that can run on the embeddings we should see how they are doing it.

yes, that was the plan, we should be able to script in one of supported languages https://opensearch.org/docs/latest/api-reference/script-apis/get-script-language/ depending on the context, starting from `painless`. Basically we need an instance of  [ScriptService ](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/script/ScriptService.java), and it's passed to the plugin class via [createComponents](https://github.com/opensearch-project/OpenSearch/blob/2.19/server/src/main/java/org/opensearch/node/Node.java#L1008).

**Comment by owaiskazi19** [comment]
@martin-gaievski @navneet1v 
I have a basic POC working for this.
1. Create a search pipeline with normalization processor with technique name as `custom`. Pass the custom normalization logic as a painless script under `parameters.script`
```json
PUT _search/pipeline/nlp-pipeline
{
    "description": "Post processor for hybrid search",
    "phase_results_processors": [
        {
            "normalization-processor": {
                "normalization": {
                    "technique": "custom",
                    "parameters": {
                        "script": "def sub_query1 = scores[0]; def sub_query2 = scores[1]; def max_score1 = max_score[0]; def max_score2 = max_score[1]; def norm_score1 = Math.min(sub_query1 / max_score1, 1.0); def norm_score2 = (sub_query2 + 1.0) / 2.0; def weight1 = 0.3; def weight2 = 0.7; ctx['result'] = weight1 * norm_score1 + weight2 * norm_score2;"
                    }
                },
                "combination": {
                    "technique": "arithmetic_mean",
                    "parameters": {
                        "weights": [
                            0.3,
                            0.4,
                            0.3
                        ]
                    }
                }
            }
        }
    ]
}
```

Here `scores` array is a list of document score of sub queries and max_score array is a list of max score of document of a sub query. We can defined some other array based on the requirement to create custom logic.

2. Create a Hybrid query with the search pipeline
```json
{
  "query": {
    "hybrid": {
      "queries": [
        {
          "match": {
            "passage_text": {
              "query": "Hi world"
            }
          }
        },
        {
          "match": {
            "passage_text": {
              "query": "planet"
            }
          }
        },
        {
          "match": {
            "passage_text": {
              "query": "Hello"
            }
          }
        }
      ]
    }
  }
}
```

Response:
```json
{
    "took": 8911,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 4,
            "relation": "eq"
        },
        "max_score": 0.52384204,
        "hits": [
            {
                "_index": "my-nlp-index",
                "_id": "1",
                "_score": 0.52384204,
                "_source": {
                    "passage_text": "Hello world",
                    "id": "s4"
                }
            },
            {
                "_index": "my-nlp-index",
                "_id": "2",
                "_score": 0.349228,
                "_source": {
                    "passage_text": "Planet",
                    "id": "s4"
                }
            },
            {
                "_index": "my-nlp-index",
                "_id": "5",
                "_score": 0.26192102,
                "_source": {
                    "passage_text": "world",
                    "id": "s4"
                }
            },
            {
                "_index": "my-nlp-index",
                "_id": "6",
                "_score": 0.26192102,
                "_source": {
                    "passage_text": "Hi",
                    "id": "s4"
                }
            }
        ]
    }
}
```

Questions:
1. Which all in built structures should we create which can help to create custom logic?

**Comment by vibrantvarun** [comment]
> {
    "took": 8911,
    "timed_out": false,
    "_shards": {

@owaiskazi19 why the response took 8911 sec?

**Comment by owaiskazi19** [comment]
> [@owaiskazi19](https://github.com/owaiskazi19) why the response took 8911 sec?

I had a debugger :)
